server/src/main/java/org/elasticsearch/rest/action/search/HttpChannelTaskHandler.java

```diff
@@ -0,0 +1,158 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.rest.action.search;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.ActionRequest;
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.admin.cluster.node.tasks.cancel.CancelTasksRequest;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.client.node.NodeClient;
+import org.elasticsearch.http.HttpChannel;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskId;
+import org.elasticsearch.tasks.TaskListener;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * This class keeps track of which tasks came in from which {@link HttpChannel}, by allowing to associate
+ * an {@link HttpChannel} with a {@link TaskId}, and also removing the link once the task is complete.
+ * Additionally, it accepts a consumer that gets called whenever an http channel gets closed, which
+ * can be used to cancel the associated task when the underlying connection gets closed.
+ */
+final class HttpChannelTaskHandler {
+    final Map<HttpChannel, CloseListener> httpChannels = new ConcurrentHashMap<>();
+
+    <Response extends ActionResponse> void execute(NodeClient client, HttpChannel httpChannel, ActionRequest request,
+                                                   ActionType<Response> actionType, ActionListener<Response> listener) {
+        //0: initial state, 1: either linked or already unlinked without being linked first, 2: first linked and then unlinked
+        //link can only be done if it's the first thing that happens. unlink will only happen if link was done first.
+        AtomicInteger link = new AtomicInteger(0);
+        Task task = client.executeLocally(actionType, request,
+            new TaskListener<>() {
+                @Override
+                public void onResponse(Task task, Response searchResponse) {
+                    unlink(task);
+                    listener.onResponse(searchResponse);
+                }
+
+                @Override
+                public void onFailure(Task task, Throwable e) {
+                    unlink(task);
+                    if (e instanceof Exception) {
+                        listener.onFailure((Exception)e);
+                    } else {
+                        //TODO should we rather throw in case of throwable instead of notifying the listener?
```

Comment:
This is called from an `ActionListener` that handles `Exception` only so getting something else than an `Exception` should not be possible. I'd just rethrow if this is a type that we don't expect.

Reply:
man, good point. We should probably fix TaskListener.

Reply:
see https://github.com/elastic/elasticsearch/pull/44946